### PR TITLE
add a new alias key for functional to non functional decompositions

### PR DESCRIPTION
**Pull request by bdhirsh**
This PR adds a new alias key for composite kernels, bringing the total count to 3. Composite kernels should be registered to this new key under the following conditions:
- The operator itself is functional (no alias annotations)
- The decomposition calls into ops that *do* have alias annotations (either mutations or views)
- The operator has a derivative formula

This alias key is similar to the `CompositeExplicitAutograd` alias key, except that the `Lazy` and `XLA` runtime keys aren't included in it (and any other future "functional" backends). That's because functional -> non-functional decompositions are not useful / kinda bad for functional backends, especially once they start using the functionalization pass (at best they introduce unnecessary overhead, at worst they can potentially cause an infinite loop with the functionalization pass).

Current key name in the PR is `CompositeExplicitAutogradWithMutations`, but I'm open to any better ideas on the name for the new key - I guess I could also use this opportunity to rename the existing composite alias keys if we think it's really worth it, although I was hoping to not have to change the existing key names :)

Other ideas:
- `CompositeExplicitAutogradNonFunctional` (better captures the fact that functional -> view decomps are also used here)
- `CompositeExplicitAutogradFunctionalToNonFunctional` (better captures the fact that the op itself should be functional, although probably way too wordy)

Note 1: in theory we could add a corresponding `CompositeImplicitAutogradWithMutations` key, but there isn't really a use case for it - it doesn't matter for functionalization, because `CompositeImplicitAutograd` kernels will *always* decompose before entering the functionalization pass.

Note 2: I was a little conflicted about what to do with mutation -> mutation decompositions (like an inplace op that decomposes to its out= variant). Semantically, I don't think it really makes sense for them to go in this key, because it doesn't matter if functional backends run them (it's only bad to *replace* a functional op with a non-functional one). We have a bunch of codegen'd `functional -> out=` and `inplace -> out=` kernels though, which I updated to get registered to the new `CompositeExplicitAutogradWithMutations` key. The `functional -> out=` kernels should definitely get registered to this key, but the `inplace -> out=` kernels probably shouldn't. But it's kind of harmless to register `inplace -> out=` kernels to the new key, and teasing that out would require adding a bunch of extra logic.

A good example of a kernel that fits this category is the  `cummax` kernel, which looks like this (and cummax has a derivative formula defined):
```
std::tuple<Tensor, Tensor> cummax(const Tensor& self, int64_t dim) {
  auto values = at::empty(self.sizes(), self.options());
  auto indices = at::empty(self.sizes(), self.options().dtype(at::kLong));
  at::cummax_out(values, indices, self, dim);
  return std::make_tuple(values, indices);
}
```

**Have all of the relevant kernels been switched over to the new key in this PR?**
No. The only ones I switched were:
- all of the codegen'd kernels
- all of the (recently added) `view_copy.out` kernels in native_functions.yaml
- A handful of one-off kernels that I noticed in native_functions.yaml

But there are probably a bunch more - I'm going to switch more over as part of integrating functionalization with LTC/XLA.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75527
* #79420
* __->__ #76358
* #78820

Differential Revision: [D35953393](https://our.internmc.facebook.com/intern/diff/D35953393)

### CODE REVIEW

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35953393).

File: aten/src/ATen/core/dispatch/OperatorEntry.cpp [hunk 1]
```diff
@@ -226,13 +229,13 @@ std::pair<const AnnotatedKernel&, const char*> OperatorEntry::computeDispatchTab
...
   //          in this case.
-  //    (2.3) Use kernel from DispatchKey::Autograd if available
+  //    (2.4) Use kernel from DispatchKey::Autograd if available
   //    The implementation of (2.2) relies on the invariant that for a given backend,
   //    `computeDispatchTableEntryWithDebug()` will be called for that backend's autograd key after the
   //    backend key. See Note [Refresh Runtime Autograd entries in dispatchTable_]
   //  (3) Use fallthrough kernel that are registered as fallback.
   // Alias Key Precedence:
-  //   CompositeExplicitAutograd > CompositeImplicitAutograd > Autograd
+  //   CompositExplicitAutogradWithMutations > CompositeExplicitAutograd > CompositeImplicitAutograd > Autograd
```
Comment (bdhirsh) [inline aten/src/ATen/core/dispatch/OperatorEntry.cpp:238]: I figured that the `CompositeExplicitAutogradWithMutations` kernels should get higher precedence, since they belong to a more specific group of dispatch keys than `CompositeExplicitAutograd`

File: torchgen/dest/register_dispatch_key.py [hunk 2]
```diff
@@ -302,7 +302,7 @@ def gen_structured(self, g: NativeFunctionsGroup) -> List[str]:
                 "Do not explicitly specify Meta dispatch key on structured "
                 "functions, they will be automatically generated for you"
             )
-        elif self.backend_index.dispatch_key == DispatchKey.CompositeExplicitAutograd:
+        elif self.backend_index.dispatch_key == DispatchKey.CompositeExplicitAutogradWithMutations:
```
Comment (bdhirsh) [inline torchgen/dest/register_dispatch_key.py:305]: All of the codegen logic for generating `functional -> out=` kernels should now be registered to the new key.

I mentioned this in the PR description - I guess technically it would be more correct to register the functional kernels to the `CompositeExplicitAutogradWithMutations` key, and the inplace kernels to the `CompositeExplicitAutograd` key, but it seems harmless to register them both to the new key.

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -695,7 +695,6 @@
   structured_delegate: atanh.out
   variants: function, method
   dispatch:
-    CompositeExplicitAutograd: atanh
```
Comment (bdhirsh) [inline aten/src/ATen/native/native_functions.yaml:698]: A bunch of these dispatch entries actually shouldn't be here, since they get generated automatically. It just didn't matter until now because they were shadowed by the codegen.

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35953393).

File: aten/src/ATen/core/dispatch/OperatorEntry.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/core/dispatch/OperatorEntry.cpp:238]: I don't follow. I could have renamed these `CompositeExplicitAutograd` and `CompositeExplicitAutogradWithoutMutations` and it seems to me like the latter is more specific!

File: aten/src/ATen/native/README.md [hunk 4]
```diff
@@ -305,6 +305,19 @@ There are also two special "generic" backends:
...
+    (2) *and* it calls internally into any mutation operators.
+    A common example of this type of kernel is for an out-of-place operator
+    that's implemented by creating an empty tensor, and calling into the corresponding out= operator.
+    We would like to distinguish between "ordinary" CompositeExplicitAutograd kernels
+    and these kernels, because some backends would not like
+    to decompose an out-of-place op into an inplace op.
+    LazyTensor + XLA are the two current examples of this - since they operate on a functional IR,
+    they would prefer to directly implement an out-of-place operator with their own kernel,
+    instead of using a decomposition that results in more mutation operators.
+
```
Comment (ezyang) [inline aten/src/ATen/native/README.md:320]: I'm looking at the description here and feeling that this is all very complicated. I know I was the one who suggested composite with mutations in the first place (and though I have forgotten what more complicated solution we were replacing with this), and I think we should still put this PR in, but we should keep our eye out for a more general approach to this problem. Maybe the stuff on decompositions and primtorch feeds in here. cc @Chillee @mruberry

File: torchgen/model.py [hunk 5]
```diff
@@ -571,11 +575,15 @@ def from_yaml(
...
-            "cannot specify both CompositeExplicitAutograd and CompositeImplicitAutograd on a single kernel; each "
+        composites_in_dispatch = [
+            d for d in dispatch if
+            d == DispatchKey.CompositeExplicitAutograd
+            or d == DispatchKey.CompositeExplicitAutogradWithMutations
+            or d == DispatchKey.CompositeImplicitAutograd]
+
+        assert len(composites_in_dispatch) <= 1, (
+            "cannot specify more than one of CompositeExplicitAutograd, CompositeExplicitAutogradWithMutations, "
+            "or CompositeImplicitAutograd on a single kernel; each "
```
Comment (ezyang) [inline torchgen/model.py:586]: I don't think this is right. I could have both CompositeExplicitAutograd and CompositeExplicitAutogradWithMutations, and under the precedence you have currently implemented this would not be a no-op because CompositeExplicitAutogradWithMutations would be applied to all the non-functional backends and CompositeExplicitAutograd would be applied to the functional backends. This might even be useful if you prefer the mutating composite in eager mode.

**Review by ezyang — APPROVED** [review]

File: aten/src/ATen/core/dispatch/OperatorEntry.cpp (hunk 1 above)
Comment (bdhirsh) [reply aten/src/ATen/core/dispatch/OperatorEntry.cpp:238]: I meant more specific in terms of "the number of dispatch keys in the alias set". `size(CompositeExplicitAutogradWithMutations)` < `size(CompositExplicitAutograd)` < `size(CompositeImplicitAutograd)`.

I guess this also shouldn't matter too much in practice, since we shouldn't really ever have an op that has both a `CompositExplicitAutogradWithMutations` and a `CompositExplicitAutograd` kernel.

File: torchgen/model.py (hunk 5 above)
Comment (bdhirsh) [reply torchgen/model.py:586]: Oh, that's a good point (although I'm planning on renaming this key to something like `CompositeExplicitAutogradWithAliasing`, so the mutation bit may not apply). Ok, it seems reasonable to allow for an operator to provide kernels for both keys.

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35953393).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35953393).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35953393).

**Comment by bdhirsh** [comment]
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35953393).
